// @acala-network/chopsticks-core/dist/cjs/blockchain/storage-layer.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    RemoteStorageLayer: function() {
        return RemoteStorageLayer;
    },
    StorageLayer: function() {
        return StorageLayer;
    },
    StorageValueKind: function() {
        return StorageValueKind;
    }
});
const _lodash = /*#__PURE__*/ _interop_require_default(require("lodash"));
const _index = require("../utils/index.js");
const _logger = require("../logger.js");
const _keycache = /*#__PURE__*/ _interop_require_default(require("../utils/key-cache.js"));
function _check_private_redeclaration(obj, privateCollection) {
    if (privateCollection.has(obj)) {
        throw new TypeError("Cannot initialize the same private elements twice on an object");
    }
}
function _class_apply_descriptor_get(receiver, descriptor) {
    if (descriptor.get) {
        return descriptor.get.call(receiver);
    }
    return descriptor.value;
}
function _class_apply_descriptor_set(receiver, descriptor, value) {
    if (descriptor.set) {
        descriptor.set.call(receiver, value);
    } else {
        if (!descriptor.writable) {
            throw new TypeError("attempted to set read only private field");
        }
        descriptor.value = value;
    }
}
function _class_extract_field_descriptor(receiver, privateMap, action) {
    if (!privateMap.has(receiver)) {
        throw new TypeError("attempted to " + action + " private field on non-instance");
    }
    return privateMap.get(receiver);
}
function _class_private_field_get(receiver, privateMap) {
    var descriptor = _class_extract_field_descriptor(receiver, privateMap, "get");
    return _class_apply_descriptor_get(receiver, descriptor);
}
function _class_private_field_init(obj, privateMap, value) {
    _check_private_redeclaration(obj, privateMap);
    privateMap.set(obj, value);
}
function _class_private_field_set(receiver, privateMap, value) {
    var descriptor = _class_extract_field_descriptor(receiver, privateMap, "set");
    _class_apply_descriptor_set(receiver, descriptor, value);
    return value;
}
function _class_private_method_get(receiver, privateSet, fn) {
    if (!privateSet.has(receiver)) {
        throw new TypeError("attempted to get private field on non-instance");
    }
    return fn;
}
function _class_private_method_init(obj, privateSet) {
    _check_private_redeclaration(obj, privateSet);
    privateSet.add(obj);
}
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
const logger = _logger.defaultLogger.child({
    name: 'layer'
});
const BATCH_SIZE = 1000;
var StorageValueKind;
(function(StorageValueKind) {
    StorageValueKind["Deleted"] = "Deleted";
    StorageValueKind["DeletedPrefix"] = "DeletedPrefix";
})(StorageValueKind || (StorageValueKind = {}));
var _api = /*#__PURE__*/ new WeakMap(), _at = /*#__PURE__*/ new WeakMap(), _db = /*#__PURE__*/ new WeakMap(), _keyCache = /*#__PURE__*/ new WeakMap(), _defaultChildKeyCache = /*#__PURE__*/ new WeakMap();
class RemoteStorageLayer {
    async get(key, _cache) {
        if (_class_private_field_get(this, _db)) {
            const res = await _class_private_field_get(this, _db).queryStorage(_class_private_field_get(this, _at), key);
            if (res) {
                return res.value ?? undefined;
            }
        }
        logger.trace({
            at: _class_private_field_get(this, _at),
            key
        }, 'RemoteStorageLayer get');
        const data = await _class_private_field_get(this, _api).getStorage(key, _class_private_field_get(this, _at));
        _class_private_field_get(this, _db)?.saveStorage(_class_private_field_get(this, _at), key, data);
        return data ?? undefined;
    }
    async foldInto(_into) {
        return this;
    }
    async fold() {}
    async getKeysPaged(prefix, pageSize, startKey) {
        if (pageSize > BATCH_SIZE) throw new Error(`pageSize must be less or equal to ${BATCH_SIZE}`);
        logger.trace({
            at: _class_private_field_get(this, _at),
            prefix,
            pageSize,
            startKey
        }, 'RemoteStorageLayer getKeysPaged');
        const isChild = (0, _index.isPrefixedChildKey)(prefix);
        const minPrefixLen = isChild ? _index.CHILD_PREFIX_LENGTH : _index.PREFIX_LENGTH;
        // can't handle keyCache without prefix
        if (prefix.length < minPrefixLen || startKey.length < minPrefixLen) {
            return _class_private_field_get(this, _api).getKeysPaged(prefix, pageSize, startKey, _class_private_field_get(this, _at));
        }
        let batchComplete = false;
        const keysPaged = [];
        while(keysPaged.length < pageSize){
            const nextKey = isChild ? await _class_private_field_get(this, _defaultChildKeyCache).next(startKey) : await _class_private_field_get(this, _keyCache).next(startKey);
            if (nextKey) {
                keysPaged.push(nextKey);
                startKey = nextKey;
                continue;
            }
            // batch fetch was completed
            if (batchComplete) {
                break;
            }
            // fetch a batch of keys
            const batch = await _class_private_field_get(this, _api).getKeysPaged(prefix, BATCH_SIZE, startKey, _class_private_field_get(this, _at));
            batchComplete = batch.length < BATCH_SIZE;
            // feed the key cache
            if (batch.length > 0) {
                if (isChild) {
                    _class_private_field_get(this, _defaultChildKeyCache).feed([
                        startKey,
                        ...batch
                    ]);
                } else {
                    _class_private_field_get(this, _keyCache).feed([
                        startKey,
                        ...batch
                    ]);
                }
            }
            if (batch.length === 0) {
                break;
            }
            if (_class_private_field_get(this, _db)) {
                // batch fetch storage values and save to db, they may be used later
                _class_private_field_get(this, _api).getStorageBatch(prefix, batch, _class_private_field_get(this, _at)).then((storage)=>{
                    for (const [key, value] of storage){
                        _class_private_field_get(this, _db).saveStorage(_class_private_field_get(this, _at), key, value);
                    }
                });
            }
        }
        return keysPaged;
    }
    constructor(api, at, db){
        _class_private_field_init(this, _api, {
            writable: true,
            value: void 0
        });
        _class_private_field_init(this, _at, {
            writable: true,
            value: void 0
        });
        _class_private_field_init(this, _db, {
            writable: true,
            value: void 0
        });
        _class_private_field_init(this, _keyCache, {
            writable: true,
            value: new _keycache.default(_index.PREFIX_LENGTH)
        });
        _class_private_field_init(this, _defaultChildKeyCache, {
            writable: true,
            value: new _keycache.default(_index.CHILD_PREFIX_LENGTH)
        });
        _class_private_field_set(this, _api, api);
        _class_private_field_set(this, _at, at);
        _class_private_field_set(this, _db, db);
    }
}
var _store = /*#__PURE__*/ new WeakMap(), _keys = /*#__PURE__*/ new WeakMap(), _deletedPrefix = /*#__PURE__*/ new WeakMap(), _parent = /*#__PURE__*/ new WeakMap(), _addKey = /*#__PURE__*/ new WeakSet(), _removeKey = /*#__PURE__*/ new WeakSet();
class StorageLayer {
    async get(key, cache) {
        if (_class_private_field_get(this, _store).has(key)) {
            return _class_private_field_get(this, _store).get(key);
        }
        if (_class_private_field_get(this, _deletedPrefix).some((dp)=>key.startsWith(dp))) {
            return "Deleted";
        }
        if (_class_private_field_get(this, _parent)) {
            const val = _class_private_field_get(this, _parent).get(key, false);
            if (cache) {
                _class_private_field_get(this, _store).set(key, val);
            }
            return val;
        }
        return undefined;
    }
    set(key, value) {
        switch(value){
            case "Deleted":
                _class_private_field_get(this, _store).set(key, "Deleted");
                _class_private_method_get(this, _removeKey, removeKey).call(this, key);
                break;
            case "DeletedPrefix":
                _class_private_field_get(this, _deletedPrefix).push(key);
                for (const k of _class_private_field_get(this, _keys)){
                    if (k.startsWith(key)) {
                        _class_private_field_get(this, _store).set(k, "Deleted");
                        _class_private_method_get(this, _removeKey, removeKey).call(this, k);
                    }
                }
                break;
            case undefined:
                _class_private_field_get(this, _store).delete(key);
                _class_private_method_get(this, _removeKey, removeKey).call(this, key);
                break;
            default:
                _class_private_field_get(this, _store).set(key, value);
                _class_private_method_get(this, _addKey, addKey).call(this, key);
                break;
        }
    }
    setAll(values) {
        if (!Array.isArray(values)) {
            values = Object.entries(values);
        }
        for (const [key, value] of values){
            this.set(key, value || "Deleted");
        }
    }
    async foldInto(into) {
        const newParent = await _class_private_field_get(this, _parent)?.foldInto(into);
        for (const deletedPrefix of _class_private_field_get(this, _deletedPrefix)){
            into.set(deletedPrefix, "DeletedPrefix");
        }
        for (const [key, value] of _class_private_field_get(this, _store)){
            into.set(key, await value);
        }
        return newParent;
    }
    async fold() {
        if (_class_private_field_get(this, _parent)) {
            _class_private_field_set(this, _parent, await _class_private_field_get(this, _parent).foldInto(this));
        }
    }
    async getKeysPaged(prefix, pageSize, startKey) {
        let parentFetchComplete = false;
        const parentFetchKeys = async (batchSize, startKey)=>{
            if (!_class_private_field_get(this, _deletedPrefix).some((dp)=>startKey.startsWith(dp))) {
                const newKeys = [];
                while(newKeys.length < batchSize){
                    const remote = await _class_private_field_get(this, _parent)?.getKeysPaged(prefix, batchSize, startKey) ?? [];
                    if (remote.length) {
                        startKey = remote[remote.length - 1];
                    }
                    for (const key of remote){
                        if (_class_private_field_get(this, _store).get(key) === "Deleted") {
                            continue;
                        }
                        if (_class_private_field_get(this, _deletedPrefix).some((dp)=>key.startsWith(dp))) {
                            continue;
                        }
                        newKeys.push(key);
                    }
                    if (remote.length < batchSize) {
                        parentFetchComplete = true;
                        break;
                    }
                }
                return newKeys;
            } else {
                parentFetchComplete = true;
                return [];
            }
        };
        const res = [];
        const foundNextKey = (key)=>{
            // make sure keys are unique and start with the prefix
            if (!res.includes(key) && key.startsWith(prefix)) {
                res.push(key);
            }
        };
        const iterLocalKeys = (prefix, startKey, includeFirst, endKey)=>{
            let idx = _class_private_field_get(this, _keys).findIndex((x)=>x.startsWith(startKey));
            if (_class_private_field_get(this, _keys)[idx] !== startKey) {
                idx = _class_private_field_get(this, _keys).findIndex((x)=>x.startsWith(prefix) && x > startKey);
                const key = _class_private_field_get(this, _keys)[idx];
                if (key) {
                    if (endKey && key >= endKey) {
                        return startKey;
                    }
                    foundNextKey(key);
                    ++idx;
                }
            }
            if (idx !== -1) {
                if (includeFirst) {
                    const key = _class_private_field_get(this, _keys)[idx];
                    if (key && key.startsWith(prefix) && key > startKey) {
                        foundNextKey(key);
                    }
                }
                while(res.length < pageSize){
                    ++idx;
                    const key = _class_private_field_get(this, _keys)[idx];
                    if (!key || !key.startsWith(prefix)) {
                        break;
                    }
                    if (endKey && key >= endKey) {
                        break;
                    }
                    foundNextKey(key);
                }
                return _lodash.default.last(res) ?? startKey;
            }
            return startKey;
        };
        if (prefix !== startKey && _class_private_field_get(this, _keys).find((x)=>x === startKey)) {
            startKey = iterLocalKeys(prefix, startKey, false);
        }
        // then iterate the parent keys
        let keys = await parentFetchKeys(pageSize - res.length, startKey);
        if (keys.length) {
            let idx = 0;
            while(res.length < pageSize){
                const key = keys[idx];
                if (!key || !key.startsWith(prefix)) {
                    if (parentFetchComplete) {
                        break;
                    } else {
                        keys = await parentFetchKeys(pageSize - res.length, _lodash.default.last(keys));
                        continue;
                    }
                }
                const keyPosition = _lodash.default.sortedIndex(_class_private_field_get(this, _keys), key);
                const localParentKey = _class_private_field_get(this, _keys)[keyPosition - 1];
                if (localParentKey < key) {
                    startKey = iterLocalKeys(prefix, startKey, false, key);
                }
                foundNextKey(key);
                ++idx;
            }
        }
        if (res.length < pageSize) {
            iterLocalKeys(prefix, startKey, prefix === startKey);
        }
        return res;
    }
    /**
   * Merge the storage layer into the given object, can be used to get sotrage diff.
   */ async mergeInto(into) {
        for (const [key, maybeValue] of _class_private_field_get(this, _store)){
            const value = await maybeValue;
            if (value === "Deleted") {
                into[key] = null;
            } else {
                into[key] = value;
            }
        }
    }
    constructor(parent){
        _class_private_method_init(this, _addKey);
        _class_private_method_init(this, _removeKey);
        _class_private_field_init(this, _store, {
            writable: true,
            value: new Map()
        });
        _class_private_field_init(this, _keys, {
            writable: true,
            value: []
        });
        _class_private_field_init(this, _deletedPrefix, {
            writable: true,
            value: []
        });
        _class_private_field_init(this, _parent, {
            writable: true,
            value: void 0
        });
        _class_private_field_set(this, _parent, parent);
    }
}
function addKey(key) {
    const idx = _lodash.default.sortedIndex(_class_private_field_get(this, _keys), key);
    const key2 = _class_private_field_get(this, _keys)[idx];
    if (key === key2) {
        return;
    }
    _class_private_field_get(this, _keys).splice(idx, 0, key);
}
function removeKey(key) {
    const idx = _lodash.default.sortedIndex(_class_private_field_get(this, _keys), key);
    const key2 = _class_private_field_get(this, _keys)[idx];
    if (key === key2) {
        _class_private_field_get(this, _keys).splice(idx, 1);
    }
}
